# Домашняя работа 10 (?? баллов). Предсказание победителя в Dota2.

В этом финальном домашнем задании вы будет работать над более-менее реальными данными - сведениями из онлайн игры dota2. Примените все знания, которые вы полусили за этот семестр. Попробуйте сгенерировать разные признаки.

По факту это перове соревнование в ML, в коттором мы предлагаем вам поучаствовать.

Изначально это задание было в курсе "Введение в машинное обучение" от ВШЭ.

## Задание

1. Скачайте данные отсюда: https://inclass.kaggle.com/c/dota-2-win-probability-prediction (потребуется регистрация на kaggle) или из папки `data/dota2`.
2. Прочитайте задание в файле `12-02-final-workshop`. Мы его не меняли и оставили таким, каким оно было в оригинальном курсе "Введение в машинное обучение".
3. Проведите всё исследование в одном jupyter ноутбуке.
5. Сохраниет решение в github и отправьте ссылку на ноутбук в eliademi.
6. Потом вы получите на проверку ноутбук сокурсника и задание на проверку. Вам нужно будет проверить всё и оформить отчет, отправив его в виде текста на почту в ответ на письмо с заданием.

In [64]:
import pandas as pd
import numpy as np
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.linear_model import RidgeClassifier, LogisticRegression
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn.preprocessing import StandardScaler, MinMaxScaler

In [65]:
features = pd.read_csv('features.csv', index_col='match_id')
features=features.fillna(0)
y = features['radiant_win']
X = features =features.drop(['duration','radiant_win','duration',
 'radiant_win',
 'tower_status_radiant',
 'tower_status_dire',
 'barracks_status_radiant',
 'barracks_status_dire'],axis=1)

In [3]:
for i in range(10, 100, 10):
    model = GradientBoostingClassifier(n_estimators=i, random_state=42)
    %time result=cross_val_score(model,X,y,scoring="roc_auc",cv = KFold(n_splits=5, random_state=42, shuffle=True))
    print(i,result,result.mean())

Wall time: 26 s
10 [0.66383799 0.66635457 0.66360048 0.66529818 0.66516222] 0.6648506879750012
Wall time: 50.6 s
20 [0.68083889 0.68272733 0.67969876 0.6834932  0.6855512 ] 0.6824618768044435
Wall time: 1min 11s
30 [0.68892093 0.68934663 0.68712298 0.69180598 0.69283583] 0.6900064710388155
Wall time: 1min 34s
40 [0.69264125 0.69335305 0.69153074 0.69586466 0.69680392] 0.6940387245121105
Wall time: 1min 56s
50 [0.69627399 0.69747879 0.69470891 0.69921915 0.69979097] 0.6974943609466162
Wall time: 2min 19s
60 [0.69904111 0.69974831 0.69741008 0.70187297 0.70252411] 0.7001193156741052
Wall time: 2min 42s
70 [0.70120426 0.7018134  0.699683   0.70399032 0.70389172] 0.7021165414535407
Wall time: 3min 5s
80 [0.70273274 0.70352284 0.70166409 0.70536378 0.70579495] 0.7038156791465111
Wall time: 3min 28s
90 [0.70428355 0.70489618 0.70314489 0.70636257 0.70710108] 0.7051576546297991


# Попробуем увеличить количество глубины деревьев

In [32]:
model = GradientBoostingClassifier(n_estimators=10, random_state=42,max_depth=10)
result=cross_val_score(model,X,y,scoring="roc_auc",cv = KFold(n_splits=5, random_state=42, shuffle=True))
result

array([0.68102708, 0.68164015, 0.67687357, 0.68110657, 0.68104133])

In [34]:

model = GradientBoostingClassifier(n_estimators=10, random_state=42,max_depth = 6)
result=cross_val_score(model,X,y,scoring="roc_auc",cv = KFold(n_splits=5, random_state=42, shuffle=True))
result

array([0.68433689, 0.68554282, 0.68026239, 0.68572485, 0.68804223])

In [39]:

model = GradientBoostingClassifier(n_estimators=1000, random_state=42,max_depth = 3)
%time result=cross_val_score(model,X,y,scoring="roc_auc",cv = KFold(n_splits=5, random_state=42, shuffle=True))
result

Wall time: 35min 14s


array([0.72392892, 0.72267035, 0.72333701, 0.72710896, 0.72504246])

In [41]:

model = GradientBoostingClassifier(n_estimators=1300, random_state=42,max_depth = 3)
%time result=cross_val_score(model,features,true_result,scoring="roc_auc",cv = KFold(n_splits=5, random_state=42, shuffle=True))
result

Wall time: 43min 56s


array([0.72513551, 0.72346132, 0.72400844, 0.72795167, 0.72549135])

# ОТЧЕТ 


Какие признаки имеют пропуски среди своих значений? Что могут означать пропуски в этих признаках (ответьте на этот вопрос для двух любых признаков)?

radiant_first_ward_time ,dire_flying_courier_time и т.д. Что значит что отсутствует время постановки 1 варда у сил света ? Очень просто - значит что она не ставили вард за первые 5 минут
Что значит что отсутствует время апа курицы у сил тьмы ? Очень просто - у сил тьмы курицу не апали в первые 5 минут,видимо задания старое,ибо сейчас в доте автоап курьера по таймингу



Как называется столбец, содержащий целевую переменную?

'radiant_win'


Как долго проводилась кросс-валидация для градиентного бустинга с 30 деревьями? Инструкцию по измерению времени можно найти ниже по тексту. Какое качество при этом получилось? Напомним, что в данном задании мы используем метрику качества AUC-ROC.

Wall time: 3min 25s,0.689188702


Имеет ли смысл использовать больше 30 деревьев в градиентном бустинге? Что бы вы предложили делать, чтобы ускорить его обучение при увеличении количества деревьев?


Да,имеет смысл,ведь это дает лучший результат. Ускорить процесс можно увеличениям числа деревьев и ограничением количества признаков

# ЛОГИСТИЧЕСКАЯ РЕГРЕССИЯ


In [4]:
list(X.columns)

['start_time',
 'lobby_type',
 'r1_hero',
 'r1_level',
 'r1_xp',
 'r1_gold',
 'r1_lh',
 'r1_kills',
 'r1_deaths',
 'r1_items',
 'r2_hero',
 'r2_level',
 'r2_xp',
 'r2_gold',
 'r2_lh',
 'r2_kills',
 'r2_deaths',
 'r2_items',
 'r3_hero',
 'r3_level',
 'r3_xp',
 'r3_gold',
 'r3_lh',
 'r3_kills',
 'r3_deaths',
 'r3_items',
 'r4_hero',
 'r4_level',
 'r4_xp',
 'r4_gold',
 'r4_lh',
 'r4_kills',
 'r4_deaths',
 'r4_items',
 'r5_hero',
 'r5_level',
 'r5_xp',
 'r5_gold',
 'r5_lh',
 'r5_kills',
 'r5_deaths',
 'r5_items',
 'd1_hero',
 'd1_level',
 'd1_xp',
 'd1_gold',
 'd1_lh',
 'd1_kills',
 'd1_deaths',
 'd1_items',
 'd2_hero',
 'd2_level',
 'd2_xp',
 'd2_gold',
 'd2_lh',
 'd2_kills',
 'd2_deaths',
 'd2_items',
 'd3_hero',
 'd3_level',
 'd3_xp',
 'd3_gold',
 'd3_lh',
 'd3_kills',
 'd3_deaths',
 'd3_items',
 'd4_hero',
 'd4_level',
 'd4_xp',
 'd4_gold',
 'd4_lh',
 'd4_kills',
 'd4_deaths',
 'd4_items',
 'd5_hero',
 'd5_level',
 'd5_xp',
 'd5_gold',
 'd5_lh',
 'd5_kills',
 'd5_deaths',
 'd5_items',


In [66]:
[ 'r1_hero',
 'r1_items',
 'lobby_type',]
drop = []
for i in range(1,6):
    drop.append('r' + str(i) + '_hero')
    drop.append('r' + str (i) + '_items')
    drop.append('d' + str(i) + '_hero')
    drop.append('d' + str (i) + '_items')
drop.append('lobby_type')

In [67]:
x = X.drop(drop,axis = 1)

In [68]:
scaler = StandardScaler()
x = scaler.fit_transform(x)

C:\conda\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\conda\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [8]:
for i in np.arange(0.1,4,0.2):
    model = LogisticRegression(random_state=42, C=i, penalty='l2',solver = 'lbfgs')
    print(cross_val_score(model, x, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc"))

[0.71323281 0.71547263 0.71508284 0.71647939 0.71760167]
[0.71323208 0.71546894 0.71507682 0.71648028 0.71759587]
[0.71323154 0.71546946 0.71507644 0.71648089 0.71759482]
[0.71323084 0.7154668  0.71507582 0.71648183 0.71759327]
[0.71323053 0.71546696 0.71507533 0.71648347 0.7175934 ]
[0.71323016 0.71546907 0.71507565 0.71648311 0.71759238]
[0.71322995 0.71546912 0.71507556 0.71648342 0.71759163]
[0.71323081 0.71546945 0.71507558 0.71648406 0.71759148]
[0.71323078 0.71546963 0.71507566 0.71648432 0.71759359]
[0.71323075 0.71546946 0.71507538 0.71648412 0.71759283]
[0.7132304  0.71546951 0.71507521 0.71648382 0.71759276]
[0.71323033 0.71546926 0.71507509 0.71648393 0.71759286]
[0.71323027 0.71546922 0.71507499 0.71648408 0.71759287]
[0.71323028 0.71546926 0.71507479 0.71648404 0.71759306]
[0.71323028 0.7154692  0.71507476 0.71648404 0.71759305]
[0.71323034 0.7154692  0.71507476 0.71648387 0.71759306]
[0.71323053 0.71546924 0.71507468 0.71648389 0.71759306]
[0.71323053 0.71546925 0.715074

In [9]:
for i in np.arange(0.1,4,0.2):
    model = LogisticRegression(random_state=42, C=i, penalty='l2',solver ='lbfgs')
    print(cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc"))

[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.51597178 0.50921324 0.51460449]
[0.51184163 0.51560696 0.515971

In [21]:
k =features['d1_hero']
N = k.unique().size 
print(N)
N = k.unique().max()
print(N)

108
112


In [69]:
# N — количество различных героев в выборке
N = 112
data = features
X_pick = np.zeros((data.shape[0], N))

for i, match_id in enumerate(data.index):
    for p in range(5):
        X_pick[i, data.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick[i, data.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1

C:\conda\lib\site-packages\ipykernel_launcher.py:8: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  
C:\conda\lib\site-packages\ipykernel_launcher.py:9: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  if __name__ == '__main__':


In [70]:
df = features
df_pick = pd.DataFrame(data=X_pick, index=df.index)
df_pick.head()

0    1    2    3    4    5    6    7    8    9   ...   102  103  \
match_id                                                   ...              
0         0.0  0.0  0.0 -1.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   
1         0.0  0.0  0.0  0.0  0.0  0.0 -1.0  0.0  0.0  0.0 ...   0.0  0.0   
2         0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   
3         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0   
4         0.0  0.0  0.0  0.0  0.0  0.0  0.0 -1.0  0.0  0.0 ...   0.0  0.0   

          104  105  106  107  108  109  110  111  
match_id                                          
0         1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 112 columns]

In [71]:
X = pd.merge(X, df_pick, left_index=True, right_index=True, how='outer')

In [72]:
X = X.drop(drop,axis = 1)

In [73]:
scaler = StandardScaler()
X = scaler.fit_transform(X)

C:\conda\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\conda\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [22]:
res = []
for i in np.arange(0.01,0.5,0.06):
    model = LogisticRegression(random_state=42, C=i, penalty='l2',solver ='lbfgs')
    res.append(cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc"))


In [16]:
for i in res:
    print(i,i.mean())

[0.74825566 0.75244183 0.74822701 0.75499765 0.7507122 ] 0.7509268697322797
[0.74825533 0.75243906 0.74822432 0.75499204 0.7507054 ] 0.7509232281599376
[0.74825602 0.75244003 0.74822321 0.75498933 0.75070392] 0.7509225015747417
[0.74825465 0.75243871 0.74822422 0.75498898 0.75070295] 0.7509219019535127
[0.74825531 0.75244081 0.74822265 0.75498823 0.75070241] 0.7509218829306203
[0.74825464 0.75244059 0.74822292 0.75498845 0.75070194] 0.7509217091678566
[0.74825465 0.75244021 0.74822114 0.75498864 0.75070172] 0.7509212728293104
[0.74825507 0.75243985 0.74822157 0.75498875 0.75070161] 0.7509213702939525
[0.74825517 0.75243976 0.74822241 0.75498872 0.75070168] 0.7509215482255692
[0.74825444 0.75243974 0.74822196 0.75498892 0.75070163] 0.7509213384700001
[0.74825455 0.75243978 0.74822189 0.75498852 0.75070144] 0.7509212367902587
[0.74825429 0.75243982 0.74822242 0.75498851 0.75070147] 0.7509213024295861
[0.74825452 0.75243972 0.74822117 0.75498861 0.75070143] 0.7509210906392645
[0.74825457 

# Отчет 2
Какое качество получилось у логистической регрессии над всеми исходными признаками? Как оно соотносится с качеством градиентного бустинга? Чем вы можете объяснить эту разницу? Быстрее ли работает логистическая регрессия по сравнению с градиентным бустингом?

0.71556599,результат похуже,чем с 1000 деревьями,но лучше чем градиентный бустинг с меньшим количеством деревьев,например 100,но логистическая регрессия значительно быстрее


Как влияет на качество логистической регрессии удаление категориальных признаков (укажите новое значение метрики качества)? Чем вы можете объяснить это изменение?


Положительно,ведь в регрессивные модели линейны,а категориальные признаки не несут в себе никакого значения,как измеримые,поэтому они лишь путают алгоритм


Сколько различных идентификаторов героев существует в данной игре?
108

Какое получилось качество при добавлении "мешка слов" по героям? Улучшилось ли оно по сравнению с предыдущим вариантом? Чем вы можете это объяснить?

качество улучшилось с 0.71556599 до 0.7509210376917823,что очень круто,целых 4 процента.Некоторые герои статистически имеют выше винрейт,поэтому то,какие герои участвует в игре очень важная информация,как категориальный признак она не несла положительную нагрузку для регрессии,но мешок слов,как бинарный признак несет в себе положительную нагрузку для регрессии

# другие методы 

# RidgeClassifier


In [24]:
solv=['auto', 'svd', 'cholesky', 'lsqr', 'sparse_cg', 'sag', 'saga']
for k in solv:
    model = RidgeClassifier(alpha = 1,solver=k,random_state=42)
    res=cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
    print(k,res,res.mean())

auto [0.74818862 0.75242903 0.74797957 0.75487812 0.75060998] 0.7508170651200416
svd [0.74818862 0.75242903 0.74797957 0.75487812 0.75060998] 0.7508170651200416
cholesky [0.74818862 0.75242903 0.74797957 0.75487812 0.75060998] 0.7508170651200416
lsqr [0.74823603 0.75261207 0.7479658  0.75509892 0.75067078] 0.7509167193246931
sparse_cg [0.74817106 0.75244081 0.74796703 0.75489773 0.7506146 ] 0.7508182463836514
sag [0.7481832  0.75242972 0.74797658 0.75487695 0.75061304] 0.7508158976551604
saga [0.74818681 0.75242998 0.74797906 0.75488013 0.75061023] 0.7508172429609895


In [26]:
model = RidgeClassifier(alpha = 1,solver='lsqr',random_state=42)
res=cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
print(res,res.mean())

0.48999999999999994 [0.74826748 0.75261403 0.74796396 0.75510867 0.75066997] 0.7509248242567688


# randomforest


In [26]:
model = RandomForestClassifier(n_estimators= 100,random_state=42)
res=cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")


In [27]:
print(res,res.mean())

[0.69019233 0.69386827 0.69261046 0.69257733 0.69769604] 0.693388886426866


In [28]:
model = RandomForestClassifier(n_estimators= 1300,random_state=42)
%time res=cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
print(res,res.mean())


Wall time: 1h 5min 43s
[0.70471013 0.70779522 0.705266   0.70650047 0.70977769] 0.7068099038933443


# BayesianRidge


In [74]:
model =  linear_model.BayesianRidge()
res =cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
res.mean(),res

(0.750871830597536,
 array([0.74821581, 0.75239536, 0.74804434, 0.75498169, 0.75072195]))

# Добавляем колонки ,произведения колонок


In [57]:
features = pd.read_csv('features.csv', index_col='match_id')
features=features.fillna(0)
y = features['radiant_win']
X = features.drop(['duration','radiant_win','duration',
 'radiant_win',
 'tower_status_radiant',
 'tower_status_dire',
 'barracks_status_radiant',
 'barracks_status_dire'],axis=1)

In [58]:
# N — количество различных героев в выборке
N = 112
data = features
X_pick = np.zeros((data.shape[0], N))

for i, match_id in enumerate(data.index):
    for p in range(5):
        X_pick[i, data.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick[i, data.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1

C:\conda\lib\site-packages\ipykernel_launcher.py:8: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  
C:\conda\lib\site-packages\ipykernel_launcher.py:9: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  if __name__ == '__main__':


In [59]:
[ 'r1_hero',
 'r1_items',
 'lobby_type',]
drop = ['lobby_type']
for i in range(1,6):
    drop.append('r' + str(i) + '_hero')
    drop.append('r' + str (i) + '_items')
    drop.append('d' + str(i) + '_hero')
    drop.append('d' + str (i) + '_items')
X = X.drop(drop,axis = 1)

In [60]:
df = features
df_pick = pd.DataFrame(data=X_pick, index=df.index)
df_pick.head()
X = pd.merge(X, df_pick, left_index=True, right_index=True, how='outer')
for p in range(1,6):
    X['r'+str(p)+'_gold_level' ] = X['r'+str(p)+'_gold' ] * X['r'+str(p)+'_level' ]
    X['d'+str(p)+'_gold_level' ] = -1 * X['d'+str(p)+'_gold' ] * X['d'+str(p)+'_level' ]
    X['r'+str(p)+'_kills_deaths' ] =  X['r'+str(p)+'_kills' ] * X['r'+str(p)+'_deaths' ]
    X['d'+str(p)+'_kills_deaths' ] = -1* X['d'+str(p)+'_kills' ] * X['d'+str(p)+'_deaths' ]

In [61]:
scaler = StandardScaler()
X = scaler.fit_transform(X)


C:\conda\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\conda\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [62]:
model = LogisticRegression(random_state=42, C=0.06, penalty='l2',solver ='lbfgs')
res = cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
print(res,res.mean())

[0.74920892 0.75386207 0.74847277 0.75594407 0.75158645] 0.7518148552953432


In [63]:
from sklearn import linear_model
model =  linear_model.BayesianRidge()
res =cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
res.mean(),res

(0.7517493767053145,
 array([0.74913538, 0.75359872, 0.74849248, 0.75589744, 0.75162286]))

# Попробуем поудалять колонки,

In [ ]:
выберем самые невлияющие на исход матча колонки,по моему опыту игры 

In [47]:
features = pd.read_csv('features.csv', index_col='match_id')
features=features.fillna(0)
y = features['radiant_win']
X = features.drop(['duration','radiant_win','duration',
 'radiant_win',
 'tower_status_radiant',
 'tower_status_dire',
 'barracks_status_radiant',
 'barracks_status_dire'],axis=1)
# N — количество различных героев в выборке
N = 112
data = features
X_pick = np.zeros((data.shape[0], N))

for i, match_id in enumerate(data.index):
    for p in range(5):
        X_pick[i, data.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick[i, data.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1
drop = ['lobby_type']
for i in range(1,6):
    drop.append('r' + str(i) + '_hero')
    drop.append('r' + str (i) + '_items')
    drop.append('d' + str(i) + '_hero')
    drop.append('d' + str (i) + '_items')
X = X.drop(drop,axis = 1)
drop = [ 'dire_courier_time' ,'first_blood_player1','first_blood_player2','radiant_bottle_time','dire_bottle_time','start_time']

X = X.drop(drop,axis = 1)
X.columns
df = features
df_pick = pd.DataFrame(data=X_pick, index=df.index)
df_pick.head()
X = pd.merge(X, df_pick, left_index=True, right_index=True, how='outer')
for p in range(1,6):
    X['r'+str(p)+'_gold_level' ] = X['r'+str(p)+'_gold' ] * X['r'+str(p)+'_level' ]
    X['d'+str(p)+'_gold_level' ] = -1 * X['d'+str(p)+'_gold' ] * X['d'+str(p)+'_level' ]
    X['r'+str(p)+'_kills_deaths' ] =  X['r'+str(p)+'_kills' ] * X['r'+str(p)+'_deaths' ]
    X['d'+str(p)+'_kills_deaths' ] = -1* X['d'+str(p)+'_kills' ] * X['d'+str(p)+'_deaths' ]

scaler = StandardScaler()
X = scaler.fit_transform(X)


C:\conda\lib\site-packages\ipykernel_launcher.py:17: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
C:\conda\lib\site-packages\ipykernel_launcher.py:18: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
C:\conda\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\conda\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [48]:
model = LogisticRegression(random_state=42, C=0.06, penalty='l2',solver ='lbfgs')
res = cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
print(res,res.mean())

[0.74895486 0.75369777 0.74816655 0.75581765 0.75128847] 0.7515850593297878


качество ухудшилось

In [49]:
features = pd.read_csv('features.csv', index_col='match_id')
features=features.fillna(0)
y = features['radiant_win']
X = features.drop(['duration','radiant_win','duration',
 'radiant_win',
 'tower_status_radiant',
 'tower_status_dire',
 'barracks_status_radiant',
 'barracks_status_dire'],axis=1)
# N — количество различных героев в выборке
N = 112
data = features
X_pick = np.zeros((data.shape[0], N))

for i, match_id in enumerate(data.index):
    for p in range(5):
        X_pick[i, data.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick[i, data.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1
drop = ['lobby_type']
for i in range(1,6):
    drop.append('r' + str(i) + '_hero')
    drop.append('r' + str (i) + '_items')
    drop.append('d' + str(i) + '_hero')
    drop.append('d' + str (i) + '_items')
X = X.drop(drop,axis = 1)
drop = [ 'dire_courier_time' ,'first_blood_player1','first_blood_player2','start_time']


X = X.drop(drop,axis = 1)
X.columns
df = features
df_pick = pd.DataFrame(data=X_pick, index=df.index)
df_pick.head()
X = pd.merge(X, df_pick, left_index=True, right_index=True, how='outer')
for p in range(1,6):
    X['r'+str(p)+'_gold_level' ] = X['r'+str(p)+'_gold' ] * X['r'+str(p)+'_level' ]
    X['d'+str(p)+'_gold_level' ] = -1 * X['d'+str(p)+'_gold' ] * X['d'+str(p)+'_level' ]
    X['r'+str(p)+'_kills_deaths' ] =  X['r'+str(p)+'_kills' ] * X['r'+str(p)+'_deaths' ]
    X['d'+str(p)+'_kills_deaths' ] = -1* X['d'+str(p)+'_kills' ] * X['d'+str(p)+'_deaths' ]

scaler = StandardScaler()
X = scaler.fit_transform(X)


C:\conda\lib\site-packages\ipykernel_launcher.py:17: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
C:\conda\lib\site-packages\ipykernel_launcher.py:18: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
C:\conda\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\conda\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [50]:
model = LogisticRegression(random_state=42, C=0.06, penalty='l2',solver ='lbfgs')
res = cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
print(res,res.mean())

[0.74897236 0.7536677  0.74811437 0.75581129 0.7512601 ] 0.7515651657944492


до сих пор,ухудшилось,надо дропать меньше признаков

In [51]:
features = pd.read_csv('features.csv', index_col='match_id')
features=features.fillna(0)
y = features['radiant_win']
X = features.drop(['duration','radiant_win','duration',
 'radiant_win',
 'tower_status_radiant',
 'tower_status_dire',
 'barracks_status_radiant',
 'barracks_status_dire'],axis=1)
# N — количество различных героев в выборке
N = 112
data = features
X_pick = np.zeros((data.shape[0], N))

for i, match_id in enumerate(data.index):
    for p in range(5):
        X_pick[i, data.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick[i, data.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1
drop = ['lobby_type']
for i in range(1,6):
    drop.append('r' + str(i) + '_hero')
    drop.append('r' + str (i) + '_items')
    drop.append('d' + str(i) + '_hero')
    drop.append('d' + str (i) + '_items')
X = X.drop(drop,axis = 1)
drop = [ 'dire_courier_time' ,'first_blood_player1']


X = X.drop(drop,axis = 1)
X.columns
df = features
df_pick = pd.DataFrame(data=X_pick, index=df.index)
df_pick.head()
X = pd.merge(X, df_pick, left_index=True, right_index=True, how='outer')
for p in range(1,6):
    X['r'+str(p)+'_gold_level' ] = X['r'+str(p)+'_gold' ] * X['r'+str(p)+'_level' ]
    X['d'+str(p)+'_gold_level' ] = -1 * X['d'+str(p)+'_gold' ] * X['d'+str(p)+'_level' ]
    X['r'+str(p)+'_kills_deaths' ] =  X['r'+str(p)+'_kills' ] * X['r'+str(p)+'_deaths' ]
    X['d'+str(p)+'_kills_deaths' ] = -1* X['d'+str(p)+'_kills' ] * X['d'+str(p)+'_deaths' ]

scaler = StandardScaler()
X = scaler.fit_transform(X)

C:\conda\lib\site-packages\ipykernel_launcher.py:17: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
C:\conda\lib\site-packages\ipykernel_launcher.py:18: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
C:\conda\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\conda\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [52]:
model = LogisticRegression(random_state=42, C=0.06, penalty='l2',solver ='lbfgs')
res = cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
print(res,res.mean())

[0.74911401 0.75384221 0.74846374 0.7558374  0.7513697 ] 0.7517254131905122


In [53]:
model =  linear_model.BayesianRidge()
res =cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
res.mean(),res

(0.7516458215136634,
 array([0.74896265, 0.75359916, 0.74846426, 0.75578403, 0.75141902]))

In [54]:
features = pd.read_csv('features.csv', index_col='match_id')
features=features.fillna(0)
y = features['radiant_win']
X = features.drop(['duration','radiant_win','duration',
 'radiant_win',
 'tower_status_radiant',
 'tower_status_dire',
 'barracks_status_radiant',
 'barracks_status_dire'],axis=1)
# N — количество различных героев в выборке
N = 112
data = features
X_pick = np.zeros((data.shape[0], N))

for i, match_id in enumerate(data.index):
    for p in range(5):
        X_pick[i, data.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick[i, data.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1
drop = ['lobby_type']
for i in range(1,6):
    drop.append('r' + str(i) + '_hero')
    drop.append('r' + str (i) + '_items')
    drop.append('d' + str(i) + '_hero')
    drop.append('d' + str (i) + '_items')
X = X.drop(drop,axis = 1)
drop = [ 'dire_courier_time' ,'first_blood_player1']


X = X.drop(drop,axis = 1)
X.columns
df = features
df_pick = pd.DataFrame(data=X_pick, index=df.index)
df_pick.head()
X = pd.merge(X, df_pick, left_index=True, right_index=True, how='outer')
for p in range(1,6):
    X['r'+str(p)+'_gold_level' ] = X['r'+str(p)+'_gold' ] * X['r'+str(p)+'_level' ]
    X['d'+str(p)+'_gold_level' ] = -1 * X['d'+str(p)+'_gold' ] * X['d'+str(p)+'_level' ]

scaler = StandardScaler()
X = scaler.fit_transform(X)

C:\conda\lib\site-packages\ipykernel_launcher.py:17: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
C:\conda\lib\site-packages\ipykernel_launcher.py:18: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
C:\conda\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\conda\lib\site-packages\sklearn\base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [55]:
model = LogisticRegression(random_state=42, C=0.06, penalty='l2',solver ='lbfgs')
res = cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
print(res,res.mean())

[0.74921762 0.75381348 0.74848352 0.75581569 0.75147748] 0.7517615571289387


In [ ]:
features = pd.read_csv('features.csv', index_col='match_id')
features=features.fillna(0)
y = features['radiant_win']
X = features.drop(['duration','radiant_win','duration',
 'radiant_win',
 'tower_status_radiant',
 'tower_status_dire',
 'barracks_status_radiant',
 'barracks_status_dire'],axis=1)
# N — количество различных героев в выборке
N = 112
data = features
X_pick = np.zeros((data.shape[0], N))

for i, match_id in enumerate(data.index):
    for p in range(5):
        X_pick[i, data.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick[i, data.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1
drop = ['lobby_type']
for i in range(1,6):
    drop.append('r' + str(i) + '_hero')
    drop.append('r' + str (i) + '_items')
    drop.append('d' + str(i) + '_hero')
    drop.append('d' + str (i) + '_items')
X = X.drop(drop,axis = 1)
drop = [ 'dire_courier_time' ,'first_blood_player1']


X = X.drop(drop,axis = 1)
X.columns
df = features
df_pick = pd.DataFrame(data=X_pick, index=df.index)
df_pick.head()
X = pd.merge(X, df_pick, left_index=True, right_index=True, how='outer')
for p in range(1,6):
    X['r'+str(p)+'_kills_deaths' ] =  X['r'+str(p)+'_kills' ] * X['r'+str(p)+'_deaths' ]
    X['d'+str(p)+'_kills_deaths' ] = -1* X['d'+str(p)+'_kills' ] * X['d'+str(p)+'_deaths' ]

scaler = StandardScaler()
X = scaler.fit_transform(X)

In [56]:
model = LogisticRegression(random_state=42, C=0.06, penalty='l2',solver ='lbfgs')
res = cross_val_score(model, X, y, cv=KFold(n_splits=5, random_state=42, shuffle=True), scoring="roc_auc")
print(res,res.mean())

[0.74921762 0.75381348 0.74848352 0.75581569 0.75147748] 0.7517615571289387


попробовал добавить лишь 1 перемножения,убрав другое,толку не особо


# Возможно мне бы стоило обрамить в функцию код,который я так долго копипастил у себя ?

# В любой непонятной ситуации стоит использовать нейросеть ?  

In [ ]:

from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
import numpy
 

def create_model():

    model = Sequential()
    model.add(Dense(194, input_dim=194, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))

    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model


model = KerasClassifier(build_fn=create_model, epochs=3, batch_size=10, verbose=1)
%time kfold = StratifiedKFold(n_splits=5, random_state=42, shuffle=True)
%time results = cross_val_score(model, X, y, cv=kfold,scoring="roc_auc")
print(results.mean())

CPU times: user 25 µs, sys: 3 µs, total: 28 µs
Wall time: 34.8 µs
Epoch 1/3
77783/77783 [==============================] - 37s 475us/step - loss: 0.6063 - acc: 0.6669
Epoch 2/3
77783/77783 [==============================] - 36s 464us/step - loss: 0.5864 - acc: 0.6853
Epoch 3/3
77783/77783 [==============================] - 36s 461us/step - loss: 0.5774 - acc: 0.6931
19447/19447 [==============================] - 3s 146us/step
Epoch 1/3
77784/77784 [==============================] - 37s 480us/step - loss: 0.6069 - acc: 0.6650
Epoch 2/3
77784/77784 [==============================] - 36s 463us/step - loss: 0.5872 - acc: 0.6838
Epoch 3/3
77784/77784 [==============================] - 36s 465us/step - loss: 0.5772 - acc: 0.6926
19446/19446 [==============================] - 3s 152us/step
Epoch 1/3
77784/77784 [==============================] - 38s 485us/step - loss: 0.6062 - acc: 0.6680
Epoch 2/3
77784/77784 [==============================] - 37s 470us/step - loss: 0.5863 - acc: 0.6855
Epoch 3/3
77784/77784 [==============================] - 37s 470us/step - loss: 0.5764 - acc: 0.6942
19446/19446 [==============================] - 3s 154us/step
Epoch 1/3
77784/77784 [==============================] - 38s 485us/step - loss: 0.6046 - acc: 0.6687
Epoch 2/3
77784/77784 [==============================] - 36s 467us/step - loss: 0.5854 - acc: 0.6854
Epoch 3/3
77784/77784 [==============================] - 36s 468us/step - loss: 0.5764 - acc: 0.6919
19446/19446 [==============================] - 3s 159us/step
Epoch 1/3
77785/77785 [==============================] - 38s 485us/step - loss: 0.6040 - acc: 0.6666
Epoch 2/3
77785/77785 [==============================] - 36s 467us/step - loss: 0.5856 - acc: 0.6874
Epoch 3/3
77785/77785 [==============================] - 36s 469us/step - loss: 0.5762 - acc: 0.6932
19445/19445 [==============================] - 3s 164us/step
CPU times: user 11min, sys: 1min 21s, total: 12min 22s
Wall time: 9min 28s
0.7434216633291358

Не самый плохой результат(код исполнялся на сервере гугл,т.к. у меня керас решил не работать),изначально использовал другую модель и выдало 0.697150825884162,в целом ужасное значения переобучения,лишь поигравшись с некоторыми параметрами что то вышло,хотя хотелось бы получить результат лучше,но тут нужно больше практиковаться